### ISSUE TITLE
Replace IFakeConfigurator with IFakeOptionsBuilder

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#461]
Currently `IFakeConfigurator` provides only one mechanism for configuring the fakes:

```
void ConfigureFake(object fake);
```

which allows users to set properties and/or set expectations via `A.CallTo`, but only after the Fake has been constructed.

**It doesn't easily allow making a fake Strict, having it call base methods, having it wrap an existing instance, having it implement additional interfaces, or giving it additional attributes.**

Replacing `IFakeConfigurator` with 

```c#
public interface IFakeOptionsBuilder
{
    /// <summary>
    /// Gets the priority of the options builder. When multiple builders that apply to
    /// the same type are registered, the one with the highest priority is used.
    /// </summary>
    /// <remarks>Negative values are reserved for use by FakeItEasy.</remarks>
    int Priority { get; }

    /// <summary>
    /// Whether or not this object can build options for a Fake of type <paramref name="type"/>.
    /// </summary>
    /// <param name="type">The type of fake to build options for.</param>
    /// <returns>
    /// <c>true</c> if this object can build options for a
    /// Fake of type <paramref name="type"/>. Otherwise <c>false</c>.
    /// </returns>
    bool CanBuildOptionsForFakeOfType(Type type);

    /// <summary>
    /// Manipulates <paramref name="options"/>, which will later be used to 
    /// create a Fake.
    /// </summary>
    /// <param name="typeOfFake">The type the fake object represents.</param>
    /// <param name="options">The fake options to manipulate.</param>
    void BuildOptions(Type typeOfFake, IFakeOptions options);
}
```

Will allow implementations to effect all the changes listed above, via methods we're used to from `A.Fake<T>(Action<IFakeOptionsBuilder<T>>)`. (The [Explicit Creation Options](https://github.com/FakeItEasy/FakeItEasy/wiki/Creating-Fakes#explicit-creation-options).)

We'd provide the `IFakeOptionsBuilder` interface as above, as well as a `FakeOptionsBuilder<TFake>` abstract base class that would provide its own `CanBuildOptionsForFakeOfType` method, which would only return true if the new Fake were to fake out a `TFake`.

When extending `FakeOptionsBuilder<TFake>`, `BuildOptions` will have access to a `IFakeOptions<TFake>`, which will provide type-safety and Intellisense, like when using the Explicit Creation Options. If the user-supplied fake options builder implements `IFakeOptionsBuilder` directly, `BuildOptions` will get an `IFakeOptions`, which would not provide compile-time type information about the Fake that's being created. (This is analogous to the trade-off between implementing `IDummyFactory` and extending `DummyFactory<T>`.

**Comment by blairconrad** [comment]
If it works, `FakeConfigurator<T>` implementations might look like this:

```csharp
protected override void ConfigureFake(IFakeOptionsBuilder<RobotRunsAmokEvent> options)
{
    options
        .Strict()
        .ConfigureFake(f =>
        {
            A.CallTo(() => f.CalculateTimestamp()).Returns(ConfiguredTimestamp);
        });
}
```

**Comment by blairconrad** [comment]
Progress continues. By commenting out one set of unit tests, I was able to get a build with only one failing spec.

The [DefaultFakeCreatorTests](https://github.com/FakeItEasy/FakeItEasy/blob/551629bffc188e2c73a82cea3ca06ca576ddcabb/Source/FakeItEasy.Tests/Creation/DefaultFakeCreatorTests.cs) give me grief, due to the changed interface. I spent a while today trying to "fix them up", but it's onerous and really a lot of the functionality has moved. I think I'd be better off introducing an earlier commit that covered the functionality in Specs (some of it already is) and then reducing the number of tests in this class.

The existing failing spec has to do with the intersection of `Wrapping` and `ConfigureFake`. 
After the change, 
[when_wrapping_an_object_with_a_fake_that_has_a_fake_configurator](https://github.com/FakeItEasy/FakeItEasy/blob/551629bffc188e2c73a82cea3ca06ca576ddcabb/Source/FakeItEasy.Specs/ConstructorWithVirtualCallsSpecs.cs#L217) fails. This is because the _all_ the `ConfigureFake`s are applied after the `Wrapping`. It should be possible to arrange things so that the configuration applied in the `A.Fake<>(o=>o.Whatever)` call overrides all the configuration applied by the `IFakeConfigurator`. 

Long story short, if you can stand to, stay tuned. I feel like we're getting there.

**Comment by blairconrad** [comment]
As is usual, working on the code raises additional issues and questions.

If we're to allow all the `FakeOptionsBuilder`actions in the `FakeConfigurator`, we need to think a little more about how the operations will interact with those specified in a `A.Fake()` call. I think we should establish a set of rules to define the interaction whenever one of these methods is called after another, whether it be from a `FakeConfigurator` and a `A.Fake()` or both of them from the same source.

Here's my proposal:

- `CallsBaseMethods` - should trump prior invocations of
  `CallsBaseMethods`, `ConfigureFake`, `Strict`, and `Wrapping`
- `ConfigureFake` - should "stack" with prior invocations, just as
  multiple `A.CallsTo` do, which would include overriding earlier
  `CallsBaseMethods`, `Strict`, and `Wrapping`
- `Implements` - should "stack" with prior invocations, ever accumulating more interfaces
- `Strict` - should trump prior invocations of `CallsBaseMethods`,
  `ConfigureFake`, and `Wrapping`
- `WithAdditionalAttributes` - should "stack" with prior invocations, ever accumulating more attributes
- `WithArgumentsForConstructor` - should trump prior invocations
- `Wrapping` - should trump prior invocations of `CallsBaseMethods`,
  `ConfigureFake`, `Strict`, and `Wrapping`

Possible issues with the interactions listed above:

1. There's no way to remove interfaces (from `Implements`) or
  attributes (from `WithAdditionalAttributes`) added by earlier
  calls. Won't bother any current users, since existing code only adds
  these during the `A.Fake` call. I think we can let this go
  for now. If it becomes an issue, users can request a
  "DoNotImplement" or "ClearInterfaces" or something to make the
  change. 
2. Currently `ConfigureFake` and `Wrapping` work so that
  `ConfigureFake` always wins, regardless of the call order (see
  `when_wrapping_an_object_and_then_configuring_a_method_called_by_constructor`
  and
  `when_configuring_a_method_called_by_constructor_and_then_wrapping_an_object`). The
  proposal above would work best if we changed that. The current specs
  are reasonable if `A.Fake()` was the only place we were using these
  methods, but if they now also apply to `FakeConfigurators`, I think
  a universal "last in wins" approach is probably best, to maintain
  consistency in the API.

Concerns? Counters?

**Comment by adamralph** [comment]
FTR, @blairconrad discussed this on gitter and I'm all for it. The proposal sounds good and I think we should go ahead.

There is an outstanding question about the naming, i.e. currently we have:

* `IFakeConfigurator.ConfigureFake(object)`
* `IFakeOptionsBuilder<T>.ConfigureFake(Action<T> action)`

As part of this change we could improve the naming to give some more distinction between the two variants of `ConfigureFake`.

**Comment by blairconrad** [comment]
Regarding the naming, I expect the new interface will have methods with these signatures (not names):

```C#
interface IFakeConfigurator
{
    int Priority { get; }
    bool CanConfigure(Type type);
    void Configure(IFakeOptionsBuilder optionsBuilder);
}
```

For the primary method, we're looking for name that describes the act of using a `IFakeOptionsBuilder` to specify options for the fake.
Oh! `SpecifyOptions`? Even `AddOptions`?

`CanConfigure` would still work there, a bit. I think I prefer it to variants such as `CanSpecifyOptionsFor` or `CanAddOptionsFor` (although `CanAddOptionsTo` may be better).

If we go that route for the method names, though, `IFakeConfigurator` doesn't apply so well. It wouldn't be so bad if we kept `CanConfigure`, maybe:

```C#
interface IFakeConfigurator
{
    int Priority { get; }
    bool CanConfigure(Type type);
    void AddOptions(IFakeOptionsBuilder optionsBuilder);
}
```

Anyone have anything better?

**Comment by blairconrad** [comment]
Now that #467 is closed, I'll prepare a PR for this. (I already have working code in the sandbox; I just have to update for the changes from #467's PR review.)

At this point, do any of the @FakeItEasy/owners have further comments on the [naming issue that @adamralph raised](https://github.com/FakeItEasy/FakeItEasy/issues/461#issuecomment-76746793)?

**Comment by adamralph** [comment]
Speaking naively (code blind), the interface looks odd.

> I'm a fake configurator and that means I have a priority, I can configure certain types and you can do *something* to me with a thing that builds fake options.

Huh?

**Comment by blairconrad** [comment]
@adamralph, does that comment apply to both versions of the interface, or just the `AddOptions`? That is, do you find Priority/CanConfigure/AddOptions worse than Priority/CanConfigure/Configure, or is it that both are horrible?

The interface pattern, of

* can I do this work?
* am I better suited to doing the work than someone else? (measured via priority), and
* do the work

matches `IDummyFactory`. Do you feel we should be looking at a different way of bundling the functionality (or completely reframing the problem) or just better names?

**Comment by blairconrad** [comment]
However things change, we're still faced with the problem of allowing users to give FakeItEasy some code that will affect the initial options of a new Fake. I do think that means the interface should include:

1. a discriminator
2. optionally a conflict-resolution mechanism in case discriminators match a type 
3. the code that will affect the initial options

Perhaps I'm out to lunch. If there's a better set of requirements, I'm keen to hear it.

For now, let's examine the interface so far:

The **discriminator** is responsible for determining whether the code should be run on any given Fake creation. And according to our API, the only information we have is the type of the Fake, so I think we end up with something that has the form
2. optionally a conflict-resolution mechanism in case two pieces of code apply to a type (This isn't strictly necessary, as we could force users to write better discriminators in step 1. For example, if they have a pieces of code to apply to all fakes of types whose name ends in "Factory" and they want to provide special code for "RobotFactory", they could build the exclusion into the discriminator of one object and build the inclusion of "RobotFactory" into the discriminator of the other. Ultimately including a priority 

To tell the truth, I find naming to be very difficult even when we're just talking about using the options builders in

```csharp
A.Fake<IFoo>(x => x.Strict());
```

what is `x`? I usually just call it `options`, even though that's a lie, since it's an `IFakeOptionsBuilder<T>`.





I'll proceed under the assumption that we don't want to complete

So, `IFakeConfigurator`, as the functionality is bundled, is really an `IFakeOptionsBuilderManipulator`

**Comment by adamralph** [comment]
I was in the middle of writing this in response to https://github.com/FakeItEasy/FakeItEasy/issues/461#issuecomment-115185544

I guess I'm referring to the general Priority/Can*DoSomething*/*DoSomething*.

`CanDoSomething` takes a `Type`, `DoSomething` takes a `IFakeOptionsBuilder`. Something seems off. `CanDoSomething` takes an object which describes an object type. Why does `DoSomething` take an object which convey the semantics of "I build fake options"?

After https://github.com/FakeItEasy/FakeItEasy/issues/461#issuecomment-115195115, I'm unclear. Does `IFakeConfigurator.DoSomething` take an `IFakeOptionsBuilder` or `IFakeOptionsBuilder<T>`?

**Comment by blairconrad** [comment]
@adamralph, I can't find the comment you referred to. I think I may have been a comment that I entered by accidentally ctrl-Entering while typing. I removed it within seconds, but maybe you were quick enough to see it?

I was off writing a long response, but I think your short questions home in on the point better than what I was going to say.

So we have a better reference, I've pushed my branch to https://github.com/blairconrad/FakeItEasy/tree/better-ifakeconfigurator.

Here's the interface from that branch:

```csharp
public interface IFakeConfigurator
{
    /// <summary>
    /// Gets the priority of the fake configurator. When multiple configurators that
    /// apply to the same type are registered, the one with the highest
    /// priority is used.
    /// </summary>
    int Priority { get; }

    /// <summary>
    /// Whether or not this object can configure a fake of type <paramref name="type"/>.
    /// </summary>
    /// <param name="type">The type of fake to configure.</param>
    /// <returns><c>true</c> if this object can configure a fake of type <paramref name="type"/>. Otherwise <c>false</c>.</returns>
    bool CanConfigureFakeOfType(Type type);

    /// <summary>
    /// Applies the configuration for the specified fake object.
    /// </summary>
    /// <param name="options">The fake object to configure.</param>
    [SuppressMessage("Microsoft.Naming", "CA1720:IdentifiersShouldNotContainTypeNames", MessageId = "object", Justification = "Fake object is a common term in FakeItEasy.")]
    void ConfigureFake(IFakeOptionsBuilder options);
}
```

with the requisite helpful abstract class the probably most people would use.

```csharp
public abstract class FakeConfigurator<T>
    : IFakeConfigurator
{
    /// <summary>
    /// Gets the priority of the fake configurator. When multiple configurators that
    /// apply to the same type are registered, the one with the highest
    /// priority is used.
    /// </summary>
    /// <remarks>The default implementation returns <c>0</c>.</remarks>
    public virtual int Priority
    {
        get { return 0; }
    }

    /// <summary>
    /// Whether or not this object can configure a fake of type <paramref name="type"/>.
    /// </summary>
    /// <param name="type">The type of fake to configure.</param>
    /// <returns><c>true</c> if <paramref name="type"/> is <typeparamref name="T"/>. Otherwise <c>false</c>.</returns>
    public bool CanConfigureFakeOfType(Type type)
    {
        return type == typeof(T);
    }

    /// <summary>
    /// Applies the configuration for the specified fake object.
    /// </summary>
    /// <param name="options">The fake object to configure.</param>
    void IFakeConfigurator.ConfigureFake(IFakeOptionsBuilder options)
    {
        Guard.AgainstNull(options, "options");

        this.ConfigureFake((IFakeOptionsBuilder<T>)options);
    }

    /// <summary>
    /// Configures the fake.
    /// </summary>
    /// <param name="options">The fake object.</param>
    protected abstract void ConfigureFake(IFakeOptionsBuilder<T> options);
}
```

So many people would just see the need to override `void ConfigureFake(IFakeOptionsBuilder<T>)`. Only if you wanted something more flexible would you need to implement the interface directly.

> CanDoSomething takes a Type, DoSomething takes a IFakeOptionsBuilder. Something seems off. CanDoSomething takes an object which describes an object type. Why does DoSomething take an object which convey the semantics of "I build fake options"?

However things change, we're still faced with the problem of allowing users to give FakeItEasy some code that will affect the initial options of a new Fake. I think that means the interface should include:

1. a discriminator, so we can decide when to manipulate the Fake's initial options, and
2. the code that will affect the initial options

At the time we're creating the Fake, the only information we have about it is its type. I think that's why `CanDoSomething` needs to take a type.
As far as the method that will affect the initial options, when users perform this task inline in their tests, they would use `A.Fake<T>(Action<IFakeOptionsBuilder<T>>)` In this case, `IFakeOptionsBuilder<T>` works because the compiler knows about `T` at that point.

We could adopt a completely novel form for configuring the Fakes and if it's good enough, push it back out to `A.Fake`, but I had not been thinking at that level. Maybe that would help, as I find "provide an action to manipulate a thing that builds options for a fake" to be a little unwieldy as it is, and hard to name. Ultimately, though, I think it's about as good a representation of what we need to do as is practical. We could remove the options builder layer, giving users access to the options directly, but I think the guidance that the builder provides is useful to avoid giving clients enough rope to hang themselves.

If we choose not to completely rejig initial fake state configuration, I think we need a thing that acts on `IFakeOptionsBuilder`s.
An obvious alternative to what I presented above would to have the interface contain a method that returns something that acts on the `IFakeOptionsBuilder`. That could give us an `IFakeOptionsBuilderActionFactory` to parallel `IDummyFactory`:

```csharp
public interface IFakeOptionsBuilderActionFactory
{
    int Priority { get; }

    bool CanCreateFakeOptionsBuilderActionFactoryFor(Type type);

    Action<IFakeOptionsBuilder> Create();
}
```

The name is a little unwieldy<sup>1</sup>, but is the whole thing moving in a better direction? 


----
<sup>1</sup>At least it wasn't `IPrioritizedDiscriminatingFakeOptionsBuilderActionFactory`<sup>2</sup>. It's like we're writing enterprise Java here.)
<sup>2</sup>…SingletonProxyImpl.

**Comment by blairconrad** [comment]
If we ultimately settle on a shape for the interface, we might be able to simplify `FakeOptionsBuilderActionFactory` by recognizing that `FakeOptions` is an internal concept, and that the most granular thing a client ever sees is a `IFakeOptionsBuilder`. That _could_ be reframed to be an outward-facing "fake options" concept, which the client can manipulate.

**Comment by blairconrad** [comment]
The more I look at things, the more I think that what users see as "fake options builders" could really just be "fake options" to them. That would alleviate _some_ of the (external) naming pressure, not just in the discoverable interface we're talking about, but when using `A.Fake<T>(Action<IFakeOptionsBuilder<T>>)` too.

If that sounds reasonable, we could produce an interface for the discoverable configurator like this:

```csharp
public interface IFakeOptionsConfigurator
{
    /// <summary>
    /// Gets the priority of the fake configurator. When multiple configurators that
    /// apply to the same faked type are registered, the one with the highest
    /// priority is used.
    /// </summary>
    int Priority { get; }

    /// <summary>
    /// Whether or not this object can configure options for a fake of type <paramref name="type"/>.
    /// </summary>
    /// <param name="type">The type of fake to configure options for.</param>
    /// <returns><c>true</c> if this object can configure options for a fake of type <paramref name="type"/>. Otherwise <c>false</c>.</returns>
    bool CanConfigureOptionsForFakeOfType(Type type);

    /// <summary>
    /// Applies the configuration to the specified fake options object.
    /// </summary>
    /// <param name="options">The fake options to configure.</param>
    void ConfigureFakeOptions(IFakeOptions options); // or Configure?
}
```

It would mean finding a new name for the internal `FakeOptions` (or perhaps merging with what was `FakeOptionsBulder`, which doesn't bother me too much, as I'm not sure the latter is pulling its weight), but if we like the new interface, I think it would be a worthwhile effort.

**Comment by blairconrad** [comment]
Any comments on the proposed interface changes? Either to the extension point or exposing `IFakeOptions` instead of a fake options builder?

**Comment by adamralph** [comment]
I've been trying to find the ~~courage~~ time to tackle your detailed analysis :stuck_out_tongue_closed_eyes: 

Will try and do so today.

**Comment by blairconrad** [comment]
Ha! Sorry. I hate to be a nag, but the psychic weight of the upcoming 2.0 release (I do _not_ want to do another hotfix) gets to me.
And sometimes I worry that the issues have been forgotten. But I know, everyone's busy. Don't put yourself out too much.

**Comment by blairconrad** [comment]
@adamralph, a potential timesaver: when you come back, take a gander at #520, which I extracted from https://github.com/FakeItEasy/FakeItEasy/issues/461#issuecomment-118212168.
I think it's a good change for the API, and will make implementing the rest of this option easier.
**And** it's a much shorter read!

**Comment by blairconrad** [comment]
I'll eventually come back to this once #529 and #536 are resolved, but thought I'd get some feedback on a different direction I'd like to take for the interface. 
Once we have renamed `IFakeOptionsBuilder` to `IFakeOptions`, all of a sudden the concept of an options builder is available to us. It _may_ be a little confusing to people who were used to working with `IFakeOptionsBuilder`s, but I doubt many people paid attention to the interface name, instead just hitting "." and letting the fluent API to its job.

Consider:

```c#
public interface IFakeOptionsBuilder
{
    int Priority { get; } // no change

    /// <summary>
    /// Whether or not this object can build options for a fake of type <paramref name="type"/>.
    /// </summary>
    /// <param name="type">The type of fake to build options for.</param>
    /// <returns><c>true</c> if this object can build options for a fake of type <paramref name="type"/>. Otherwise <c>false</c>.</returns>
    bool CanBuildOptionsForFakeOfType(Type type); // or simply CanBuildOptionsFor

    /// <summary>
    /// Manipulates <paramref name="options"/>, which will later be used to 
    /// create a Fake.
    /// </summary>
    /// <param name="options">The fake options to configure.</param>
    void BuildOptions(IFakeOptions options);
}
```

**Comment by adamralph** [comment]
@blairconrad 

> I'll eventually come back to this once #529 and #536 are resolved

They are resolved :smile:.

I'd like to give you some feedback for some of the proposals above, but I fear I'm rather lost in the lengthy history of this issue.

**Comment by blairconrad** [comment]
Oh, I'll just send a PR today or tomorrow anyhow. Pick that apart.

**Comment by blairconrad** [comment]
I'm back.

Short version is that I propose creating

```c#
public interface IFakeOptionsBuilder
{
    /// <summary>
    /// Gets the priority of the options builder. When multiple builders that apply to
    /// the same type are registered, the one with the highest priority is used.
    /// </summary>
    /// <remarks>Negative values are reserved for use by FakeItEasy.</remarks>
    int Priority { get; }

    /// <summary>
    /// Whether or not this object can build options for a Fake of type <paramref name="type"/>.
    /// </summary>
    /// <param name="type">The type of fake to build options for.</param>
    /// <returns>
    /// <c>true</c> if this object can build options for a
    /// Fake of type <paramref name="type"/>. Otherwise <c>false</c>.
    /// </returns>
    bool CanBuildOptionsForFakeOfType(Type type);

    /// <summary>
    /// Manipulates <paramref name="options"/>, which will later be used to 
    /// create a Fake.
    /// </summary>
    /// <param name="typeOfFake">The type the fake object represents.</param>
    /// <param name="options">The fake options to manipulate.</param>
    void BuildOptions(Type typeOfFake, IFakeOptions options);
}
```

and a corresponding `IFakeOptionsBuilder<TFake>` to replace `IFakeConfigurator`.
`IFakeOptions` will let you do everything you can do with `IFakeOptions<TFake>` (but without the compile-time type safety).

**Comment by adamralph** [comment]
@blairconrad can you give an example of what this will allow the user to do? That's what I'm missing right now. Or is that somewhere in the tests in the #574 changeset?

**Comment by blairconrad** [comment]
Currently, the only configuration that can be performed via the `IFakeConfigurator` is to manipulate an already-created fake via the `ConfigureFake(object)` mechanism. So it's mostly being able to set properties or call methods on the fake or to use `A.CallTo` to configure it.

With this change, anything you can do with an `IFakeOptions<T>` can be done to the fake (even before it's created):

- `ConfigureFake`, as before
- `WithAgumentsForCosntructor`
- `WithAdditionalAttributes`
- `Wrapping`
- `Implements`

The new mechanism would have exactly the same power as the

```c#
public static T Fake<T>(Action<IFakeOptions<T>> optionsBuilder)
```
mechanism.

And by using `CanBuildOptionsForFakeOfType` and the new `IFakeOptions`, we can apply these configuration options to many many kinds of fake. For example, with one concrete FakeOptionsBuilder, users could have _all_ fakes be strict.
Or any fake that ends with "Factory" could implement a given interface, or every "*Controller" could have a service bus supplied to its constructor. Or something.

https://github.com/FakeItEasy/FakeItEasy/pull/574/files#diff-a182267575c9320c630f864a6aec4bc9 is chock full of examples.

**Comment by adamralph** [comment]
OK, I think I'm getting it, so it would allow us to add APIs which allow doing things to fakes after creation, which can currently only be done during creation? E.g.

```C#
A.CallTo(myFake).IsStrict();
```

**Comment by blairconrad** [comment]
Rather the opposite, I think. I seem to be doing a terrible job of explaining things. :disappointed: 

Consider the existing discoverable `IFakeConfigurator` extension point. A selected `IFakeConfigurator` is passed a fully-constructed Fake and then gets the option of modifying it via its `ConfigureFake` method. But since the Fake is already constructed, we can't (for example) cause it to implement additional interfaces or have extra attributes attached to it or wrap another object.
This is rather limited, especially when we consider that all of these operations _are_ available as [Explicit Creation Options](https://github.com/FakeItEasy/FakeItEasy/wiki/Creating-Fakes#explicit-creation-options)  via the `public static T Fake<T>(Action<IFakeOptions<T>> optionsBuilder)` method on `A`.
My change would replace `IFakeConfigurator` with `IFakeOptionsBuilder` and implementations of the latter would enable "Implicit Creation Options" which are as rich as the Explicit Creation Options: all creation options would be available and would be applied _before_ fake creation, just as when using `A.Fake(<Action<IFakeOptions<T>>)`.

**Comment by adamralph** [comment]
Ah, I see. In the interest of making this new feature easily grok-able by users, shall we change the title to something like "Replace IFakeConfigurator with IFakeOptionsBuilder" and perhaps prune down the description to focus on describing why an `IFakeOptionsBuilder` will be better?

**Comment by blairconrad** [comment]
Excellent idea. The description did not keep up with the discussion in the comments. Take a gander.

**Comment by adamralph** [comment]
:+1: looks good

**Comment by thomaslevesque** [comment]
Just to make sure I understand... If I do `A.Fake<IFoo>()` without specifying options, and there is a fake configurator (or options builder, whatever it's called) for `IFoo`, then the options specified by the fake configurator will be used, right? Is there a way to override that? For instance, if the fake configurator says that the fake must be strict, will I still be able to create a non-strict fake?

**Comment by blairconrad** [comment]
Actually, the fake configurator (or options builder, in my hopeful future) would be applied regardless to any of the fakes.
Then if you specify options either via `A.Fake<IFoo>(o => …)` or by subsequent `A.CallTo`s, they will take precedence - last in wins, kind of thing.
As far as making something non-strict, just as if it were specified via `o => o.Strict()`, you'd have to add overriding configuration.

**Comment by thomaslevesque** [comment]
> As far as making something non-strict, just as if it were specified via o => o.Strict(), you'd have to add overriding configuration.

Do you mean adding a `NonStrict()` method ?

**Comment by blairconrad** [comment]
I just meant overriding `A.CallTo()`s. I'd prefer not to add a `NonStrict`, since non-strict is the default. If people want fakes to not be strict, just don't make them strict. If they want some of their objects to be strict, turn those ones strict as needed.

**Comment by blairconrad** [comment]
Thanks! I'll update the release notes in the morning if nobody else has.
@thomaslevesque I think you wanted this in the next beta, and @adamralph and I were happy to have a release any time.
If I've not muddled that, shall we release? It may be the last one before 2.0.0 is "feature complete" (I can dream).

**Comment by blairconrad** [comment]
If we want to release, I should be around starting at about 11:00 UTC for an hour or so to chat about unclear release processes, @thomaslevesque.

**Comment by blairconrad** [comment]
Of course, one of the things needed for a release is good release notes. I updated them.

**Comment by thomaslevesque** [comment]
Yes, we can release whenever you want.

> If we want to release, I should be around starting at about 11:00 UTC for an hour or so to chat about unclear release processes, @thomaslevesque.

I won't be available before 19:00 UTC... too much to do at work

**Comment by blairconrad** [comment]
Oh, sure. I didn't mean to suggest you should start at 11:00 or anything. Just indicating when I start being around. Hoot me whenever you like. Of course, I'll still be at work at 19:00, but that shouldn't preclude some chatting. :-)

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
